File: backend/receiver.js
/**
 * Rahat Sentinel — Node.js Disaster Response Simulation Pipeline
 *
 * Connects to the Sentinel FastAPI WebSocket or runs entirely offline via --test.
 * Simulates the full journey: Satellite Detection → AI Grid Analysis →
 * Network Failure → BLE Mesh Propagation → Real-World Rescue deployment.
 *
 * Usage:
 *   node receiver.js            # live mode
 *   node receiver.js --test     # demo mode
 */

const WebSocket = require("ws");

// ─── Config ──────────────────────────────────────────────────────────
const WS_URL = "ws://localhost:8000/ws/alerts";
const RECONNECT_DELAY_MS = 2000;
const SIMULATE_INTERVAL_MS = 12000;
const MAX_RETRY_LOG = 3; 

// Region Bounding Box (Bhilai-Durg)
const BOUNDS = {
  minLat: 21.15,
  maxLat: 21.30,
  minLon: 81.25,
  maxLon: 81.45,
};
const GRID_SIZE = 20;

// Explanation reasons for the "AI Interpretability" log
const AI_REASONS = [
  "High surface temperature anomaly detected",
  "Vegetation index drop + sudden heat spike",
  "Rapid thermal expansion matching wildfire signature",
  "Atmospheric smoke density exceeds threshold",
  "Unseasonal thermal cluster detected near built-up area"
];

// ─── ANSI Colours ────────────────────────────────────────────────────
const C = {
  reset: "\x1b[0m",
  bold: "\x1b[1m",
  dim: "\x1b[2m",
  red: "\x1b[91m",
  yellow: "\x1b[93m",
  green: "\x1b[92m",
  cyan: "\x1b[96m",
  magenta: "\x1b[95m",
  white: "\x1b[97m",
  bgRed: "\x1b[41m",
};

// ─── Helpers ─────────────────────────────────────────────────────────

function ts() {
  return new Date().toLocaleTimeString("en-GB", { hour12: false });
}

function log(msg) {
  console.log(`  ${C.dim}[${ts()}]${C.reset}  ${msg}`);
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ─── Grid Logic ──────────────────────────────────────────────────────

function getGridCell(lat, lon) {
  const cLat = Math.max(BOUNDS.minLat, Math.min(BOUNDS.maxLat, lat));
  const cLon = Math.max(BOUNDS.minLon, Math.min(BOUNDS.maxLon, lon));
  const row = Math.floor(((BOUNDS.maxLat - cLat) / (BOUNDS.maxLat - BOUNDS.minLat)) * (GRID_SIZE - 1));
  const col = Math.floor(((cLon - BOUNDS.minLon) / (BOUNDS.maxLon - BOUNDS.minLon)) * (GRID_SIZE - 1));
  return { row, col, lat: cLat, lon: cLon };
}

function getGridCoords(row, col) {
  const latDelta = (BOUNDS.maxLat - BOUNDS.minLat) / (GRID_SIZE - 1);
  const lonDelta = (BOUNDS.maxLon - BOUNDS.minLon) / (GRID_SIZE - 1);
  return { lat: BOUNDS.maxLat - row * latDelta, lon: BOUNDS.minLon + col * lonDelta };
}

// ─── BLE Payload Encoder ─────────────────────────────────────────────

function encodeBLE(data, grid) {
  const compact = {
    r: "Bhilai-Durg",
    s: (typeof data.severity === "number" ? data.severity : 0).toFixed(3),
    t: data.timestamp || new Date().toISOString(),
    gx: grid.col,
    gy: grid.row,
  };
  return Buffer.from(JSON.stringify(compact)).toString("base64");
}

// ─── Alert Processing Pipeline ───────────────────────────────────────

let processing = false;

async function processAlert(data, isClustered = false) {
  if (processing) return; 
  processing = true;

  try {
    const severity = data && typeof data.severity === "number" ? data.severity : 0;
    const timestamp = (data && data.timestamp) || new Date().toISOString();

    let grid;
    if (data && typeof data.lat === "number" && typeof data.lon === "number" && !isNaN(data.lat)) {
      grid = getGridCell(data.lat, data.lon);
    } else if (data && typeof data.grid_row === "number" && typeof data.grid_col === "number") {
      grid = { row: data.grid_row, col: data.grid_col, ...getGridCoords(data.grid_row, data.grid_col) };
    } else {
      grid = getGridCell((BOUNDS.minLat + BOUNDS.maxLat) / 2, (BOUNDS.minLon + BOUNDS.maxLon) / 2);
    }

    const sc = severity >= 0.8 ? C.red : severity >= 0.5 ? C.yellow : C.green;
    const reason = AI_REASONS[Math.floor(Math.random() * AI_REASONS.length)];

    // 1. Initial Timeline
    console.log();
    log(`${C.cyan}🛰️  Satellite anomaly detected${C.reset}`);
    await sleep(600);
    log(`${C.magenta}🤖 AI model flagged risk${C.reset}`);
    await sleep(600);
    log(`${C.yellow}🧠 Analyzing spatial patterns...${C.reset}`);
    await sleep(400);

    if (isClustered) {
      console.log(`\n  ${C.bold}${C.red}⚠️ Multiple hotspots detected across grid${C.reset}`);
      await sleep(600);
    }

    // 2. Alert Generation
    console.log();
    console.log(`  ${sc}${C.bold}${"=".repeat(30)}${C.reset}`);
    console.log(`  ${sc}${C.bold}🚨 EMERGENCY ALERT TRIGGERED${C.reset}`);
    console.log(`  ${sc}${C.bold}${"=".repeat(30)}${C.reset}`);
    console.log();
    console.log(`  ${C.dim}🛰️${C.reset} Region:     ${C.cyan}Bhilai-Durg${C.reset}`);
    console.log(`  ${C.dim}📍${C.reset} Grid Cell:  [${grid.row}, ${grid.col}]   ${C.dim}(20x20 matrix)${C.reset}`);
    console.log(`  ${C.dim}🌡️${C.reset} Severity:   ${sc}${severity.toFixed(3)}${C.reset}`);
    console.log(`  ${C.dim}🕒${C.reset} Time:       ${timestamp}`);
    console.log(`  ${C.dim}📌${C.reset} Location:   ${grid.lat.toFixed(4)}°N, ${grid.lon.toFixed(4)}°E`);
    console.log();
    log(`🧠 ${C.magenta}Reason: ${C.reset}${C.dim}${reason}${C.reset}`);
    console.log();

    if (severity > 0.8) {
      console.log(`  ${C.bgRed}${C.white}${C.bold} 🔥 CRITICAL ALERT — IMMEDIATE ACTION REQUIRED ${C.reset}`);
      console.log();
    }
    
    if (isClustered && severity > 0.85) {
      await sleep(400);
      console.log(`  ${C.bgRed}${C.white}${C.bold} 🚨 EMERGENCY LEVEL ESCALATED — MULTIPLE CRITICAL ZONES 🚨 ${C.reset}`);
      console.log();
    }

    // 3. Network Failure 
    await sleep(800);
    log(`📴 ${C.red}Network failure detected in sector [${grid.row}, ${grid.col}]${C.reset}`);
    await sleep(600);
    log(`🔁 ${C.cyan}Switching to offline BLE mesh...${C.reset}`);
    console.log();

    // 4. BLE Payload & Propagation
    let payload = "";
    try {
      payload = encodeBLE(data, grid);
      log(`📡 ${C.magenta}BLE PAYLOAD READY: ${C.reset}${C.dim}${payload}${C.reset}`);
    } catch (err) {
      log(`${C.red}✖ BLE encode failed: ${err.message}${C.reset}`);
    }

    await sleep(1000);
    log(`${C.cyan}📡 Broadcasting via BLE mesh...${C.reset}`);
    console.log();
    
    await sleep(1000);
    log(`  ${C.green}📲 Device 2 received alert${C.reset}  ${C.dim}(hop 1)${C.reset}`);
    await sleep(800);
    log(`  ${C.green}📲 Device 3 received alert${C.reset}  ${C.dim}(hop 2)${C.reset}`);
    await sleep(800);
    log(`  ${C.green}📲 Device 4 received alert${C.reset}  ${C.dim}(hop 3)${C.reset}`);
    console.log();

    // 5. Real World Impact
    await sleep(800);
    log(`🚑 ${C.yellow}Rescue teams notified at staging area${C.reset}`);
    await sleep(600);
    log(`📢 ${C.yellow}Nearby civilians alerted (150m radius)${C.reset}`);
    await sleep(600);
    log(`🛟 ${C.green}${C.bold}Emergency response initiated${C.reset}`);
    console.log();

  } finally {
    processing = false;
  }
}

// ─── WebSocket Message Handler ───────────────────────────────────────

function handleMessage(raw) {
  let msg;
  try {
    msg = JSON.parse(raw);
  } catch {
    return; // graceful skip
  }

  if (msg.event && msg.data !== undefined) {
    if (msg.event === "new_alert" && typeof msg.data === "object" && !Array.isArray(msg.data)) {
      processAlert(msg.data);
    }
    return;
  }

  if (msg.region || msg.severity !== undefined) {
    processAlert(msg);
  }
}

// ─── WebSocket Connection ────────────────────────────────────────────

let retryCount = 0;

function connect() {
  log(`🔌 Connecting to WebSocket...`);

  let ws;
  try {
    ws = new WebSocket(WS_URL);
  } catch (err) {
    scheduleReconnect();
    return;
  }

  ws.on("open", () => {
    retryCount = 0;
    log(`✅ ${C.green}Connected to server${C.reset}`);

    const pingInterval = setInterval(() => {
      if (ws.readyState === WebSocket.OPEN) ws.send("ping");
    }, 25000);

    ws.on("close", () => {
      clearInterval(pingInterval);
      log(`${C.yellow}⚠ Disconnected from server${C.reset}`);
      scheduleReconnect();
    });
  });

  ws.on("message", (data) => handleMessage(data.toString()));
  ws.on("error", () => {
    if (retryCount === 0) log(`${C.yellow}⚠ WebSocket connection failed${C.reset}`);
  });
}

function scheduleReconnect() {
  retryCount++;
  if (retryCount === 1) {
    log(`${C.yellow}⚠ Server unavailable — switching to test mode${C.reset}`);
    startTestMode();
    return;
  }
  if (retryCount <= MAX_RETRY_LOG) {
    log(`🔄 Reconnecting in ${RECONNECT_DELAY_MS / 1000}s...`);
  }
  setTimeout(connect, RECONNECT_DELAY_MS);
}

// ─── Test / Demo Mode (Realistic Anomalies) ──────────────────────────

let demoInterval;
let clusterCenter = null;

function generateSimulatedAlert() {
  // 1/5 chance to trigger a hotspot cluster
  const isClustered = Math.random() > 0.8;
  
  if (isClustered && !clusterCenter) {
    clusterCenter = {
      row: Math.floor(Math.random() * (GRID_SIZE - 4)) + 2,
      col: Math.floor(Math.random() * (GRID_SIZE - 4)) + 2,
    };
  } else if (!isClustered) {
    clusterCenter = null;
  }

  let row, col, severity;

  if (clusterCenter) {
    // Generate near cluster center, high severity
    row = clusterCenter.row + Math.floor(Math.random() * 3) - 1;
    col = clusterCenter.col + Math.floor(Math.random() * 3) - 1;
    severity = +(Math.random() * 0.15 + 0.85).toFixed(3); // 0.85 to 1.00
  } else {
    // Generate randomly, lower severity
    row = Math.floor(Math.random() * GRID_SIZE);
    col = Math.floor(Math.random() * GRID_SIZE);
    severity = +(Math.random() * 0.4 + 0.4).toFixed(3); // 0.40 to 0.80
  }
  
  return { data: {
    region: "Bhilai-Durg",
    severity,
    timestamp: new Date().toISOString(),
    grid_row: row,
    grid_col: col
  }, isClustered: !!clusterCenter };
}

function startTestMode() {
  if (demoInterval) return;

  console.log();
  console.log(`  ${C.bold}${C.yellow}${"─".repeat(52)}${C.reset}`);
  console.log(`  ${C.bold}${C.yellow}  ⚡  TEST MODE — simulating full pipeline${C.reset}`);
  console.log(`  ${C.bold}${C.yellow}${"─".repeat(52)}${C.reset}`);

  // Fire one immediately, then loop
  const initial = generateSimulatedAlert();
  processAlert(initial.data, initial.isClustered);
  
  demoInterval = setInterval(() => {
    const next = generateSimulatedAlert();
    processAlert(next.data, next.isClustered);
  }, SIMULATE_INTERVAL_MS); // Spaced out to let the 11-second timeline finish printing
}

// ─── Entry Point ─────────────────────────────────────────────────────

function main() {
  console.log();
  console.log(`  ${C.bold}${"═".repeat(60)}${C.reset}`);
  console.log(`  ${C.bold}${C.cyan}  Rahat Sentinel — Disaster Response Pipeline Emulator${C.reset}`);
  console.log(`  ${C.bold}${"═".repeat(60)}${C.reset}`);
  console.log(`  ${C.dim}Target:${C.reset}  ${WS_URL}`);
  console.log();

  const testMode = process.argv.includes("--test");

  if (testMode) {
    startTestMode();
  } else {
    connect();
  }
}

main();

File: backend/receiver.py
"""
Rahat Sentinel — WebSocket Alert Receiver & BLE Encoder.

Connects to the Sentinel FastAPI WebSocket, receives live alerts,
encodes each into a 47-byte BLE payload, and logs everything
with colour-coded, timestamped output for demo clarity.

Usage:
    cd backend/
    python -m receiver          # connects to ws://localhost:8000/ws/alerts
    python -m receiver --test   # skips WS, simulates alerts every 2 s
"""

from __future__ import annotations

import argparse
import asyncio
import json
import logging
import random
import struct
import sys
import uuid
from datetime import datetime, timezone
from typing import Any

try:
    import websockets
    from websockets.exceptions import (
        ConnectionClosed,
        InvalidURI,
        WebSocketException,
    )
except ImportError:
    print("✖  'websockets' package not found. Install with: pip install websockets")
    sys.exit(1)

from app.ble_codec import pack_alert, unpack_alert, PAYLOAD_SIZE
from app.models import SentinelAlert

# ---------------------------------------------------------------------------
# ANSI colour helpers
# ---------------------------------------------------------------------------

class C:
    """ANSI escape codes for coloured terminal output."""
    RESET   = "\033[0m"
    BOLD    = "\033[1m"
    DIM     = "\033[2m"
    RED     = "\033[91m"
    YELLOW  = "\033[93m"
    GREEN   = "\033[92m"
    CYAN    = "\033[96m"
    MAGENTA = "\033[95m"
    WHITE   = "\033[97m"
    BG_RED  = "\033[41m"
    BG_YEL  = "\033[43m"
    BG_GRN  = "\033[42m"


def _severity_colour(severity: int) -> str:
    """Return an ANSI colour based on severity (0-255)."""
    if severity >= 180:
        return C.RED
    if severity >= 100:
        return C.YELLOW
    return C.GREEN


def _ts() -> str:
    """Compact local timestamp for log lines."""
    return datetime.now().strftime("%H:%M:%S")


# ---------------------------------------------------------------------------
# Core alert processing
# ---------------------------------------------------------------------------

def process_alert(alert_data: dict[str, Any]) -> None:
    """
    Process a single alert dict:
      1. Build a SentinelAlert model
      2. Log formatted alert details
      3. Encode to BLE binary
      4. Log the hex payload
    """
    try:
        alert = SentinelAlert(**alert_data)
    except Exception as exc:
        print(f"  {C.DIM}[{_ts()}]{C.RESET}  {C.YELLOW}⚠  Skipping malformed alert: {exc}{C.RESET}")
        return

    sev_colour = _severity_colour(alert.severity)

    # ── Formatted alert log ──────────────────────────────────────────
    print()
    print(f"  {C.DIM}[{_ts()}]{C.RESET}  {sev_colour}{C.BOLD}🚨 ALERT RECEIVED{C.RESET}")
    print(f"  {C.DIM}├─{C.RESET} Region:    {C.CYAN}{alert.district}{C.RESET}  (id {alert.region_id})")
    print(f"  {C.DIM}├─{C.RESET} Severity:  {sev_colour}{alert.severity}/255{C.RESET}  "
          f"(risk {alert.risk_score:.3f})")
    print(f"  {C.DIM}├─{C.RESET} Time:      {alert.timestamp}")
    print(f"  {C.DIM}├─{C.RESET} Center:    {alert.center.lat:.4f}°N, {alert.center.lon:.4f}°E")
    print(f"  {C.DIM}└─{C.RESET} GradCAM:   {alert.gradcam_hash}")

    # ── BLE encoding ─────────────────────────────────────────────────
    try:
        payload = pack_alert(alert)
        hex_str = payload.hex()
        print(f"  {C.DIM}[{_ts()}]{C.RESET}  {C.MAGENTA}📡 BLE PAYLOAD READY{C.RESET}  "
              f"({len(payload)}B)  {C.DIM}{hex_str}{C.RESET}")
    except Exception as exc:
        print(f"  {C.DIM}[{_ts()}]{C.RESET}  {C.RED}✖  BLE encode failed: {exc}{C.RESET}")


# ---------------------------------------------------------------------------
# WebSocket listener
# ---------------------------------------------------------------------------

WS_URL = "ws://localhost:8000/ws/alerts"
RETRY_DELAY = 3  # seconds


async def _handle_message(raw: str) -> None:
    """Parse a WebSocketMessage envelope and process alerts inside."""
    try:
        msg = json.loads(raw)
    except json.JSONDecodeError as exc:
        print(f"  {C.DIM}[{_ts()}]{C.RESET}  {C.YELLOW}⚠  Bad JSON, skipping: {exc}{C.RESET}")
        return

    event = msg.get("event", "")
    data = msg.get("data")

    if event == "new_alert" and isinstance(data, dict):
        process_alert(data)

    elif event == "history" and isinstance(data, list):
        print(f"  {C.DIM}[{_ts()}]{C.RESET}  {C.CYAN}📜 Received {len(data)} cached alert(s){C.RESET}")
        for alert_data in data:
            process_alert(alert_data)

    elif event == "heartbeat":
        print(f"  {C.DIM}[{_ts()}]{C.RESET}  {C.DIM}💓 heartbeat{C.RESET}")

    else:
        print(f"  {C.DIM}[{_ts()}]{C.RESET}  {C.DIM}ℹ  Unknown event: {event}{C.RESET}")


async def ws_listener() -> None:
    """Connect to the WS server and listen forever with auto-reconnect."""
    print(f"\n  {C.BOLD}{'═' * 52}{C.RESET}")
    print(f"  {C.BOLD}{C.CYAN}  Rahat Sentinel — Alert Receiver & BLE Encoder{C.RESET}")
    print(f"  {C.BOLD}{'═' * 52}{C.RESET}")
    print(f"  {C.DIM}Target:{C.RESET}  {WS_URL}")
    print()

    first_attempt = True

    while True:
        try:
            print(f"  {C.DIM}[{_ts()}]{C.RESET}  Connecting to WebSocket …")
            async with websockets.connect(WS_URL) as ws:
                first_attempt = False
                print(f"  {C.DIM}[{_ts()}]{C.RESET}  {C.GREEN}✔  Connected!{C.RESET}")

                # Send periodic pings to keep connection alive
                async def _ping_loop():
                    while True:
                        await asyncio.sleep(25)
                        try:
                            await ws.send("ping")
                        except Exception:
                            break

                ping_task = asyncio.create_task(_ping_loop())

                try:
                    async for raw_msg in ws:
                        await _handle_message(raw_msg)
                finally:
                    ping_task.cancel()

        except (ConnectionClosed, ConnectionRefusedError, OSError) as exc:
            label = type(exc).__name__
            print(f"  {C.DIM}[{_ts()}]{C.RESET}  {C.YELLOW}⚠  {label} — ", end="")
            if first_attempt:
                print(f"server unavailable, switching to test mode{C.RESET}")
                await _fallback_test_loop()
                return
            print(f"retrying in {RETRY_DELAY}s …{C.RESET}")
            await asyncio.sleep(RETRY_DELAY)

        except WebSocketException as exc:
            print(f"  {C.DIM}[{_ts()}]{C.RESET}  {C.RED}✖  WebSocket error: {exc}{C.RESET}")
            await asyncio.sleep(RETRY_DELAY)

        except KeyboardInterrupt:
            print(f"\n  {C.DIM}[{_ts()}]{C.RESET}  {C.CYAN}Shutting down …{C.RESET}")
            return


# ---------------------------------------------------------------------------
# Fallback test mode — simulates alerts when WS is unavailable
# ---------------------------------------------------------------------------

DEMO_DISTRICTS = [
    "NIT Delhi", "North Delhi", "Civil Lines", "Rohini",
    "Model Town", "Pitampura", "Burari", "Shalimar Bagh",
]

# NIT Delhi center: 28.7453°N, 77.1157°E
_NIT_LAT, _NIT_LON = 28.7453, 77.1157


def _random_alert_data() -> dict[str, Any]:
    """Generate a fake SentinelAlert-shaped dict for testing."""
    risk = round(random.uniform(0.3, 1.0), 3)
    return {
        "id": uuid.uuid4().hex[:12],
        "risk_score": risk,
        "cells": [[round(random.uniform(0, 1), 2) for _ in range(20)] for _ in range(20)],
        "district": random.choice(DEMO_DISTRICTS),
        "region_id": random.randint(1, 50),
        "severity": int(risk * 255),
        "gradcam_hash": uuid.uuid4().hex[:16],
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "center": {
            "lat": round(_NIT_LAT + random.uniform(-0.05, 0.05), 4),
            "lon": round(_NIT_LON + random.uniform(-0.05, 0.05), 4),
        },
    }


async def _fallback_test_loop() -> None:
    """Emit simulated alerts every 2 seconds."""
    print()
    print(f"  {C.BOLD}{C.YELLOW}{'─' * 52}{C.RESET}")
    print(f"  {C.BOLD}{C.YELLOW}  ⚡  TEST MODE — simulating alerts every 2 s{C.RESET}")
    print(f"  {C.BOLD}{C.YELLOW}{'─' * 52}{C.RESET}")
    print()

    try:
        while True:
            data = _random_alert_data()
            process_alert(data)
            await asyncio.sleep(2)
    except KeyboardInterrupt:
        print(f"\n  {C.DIM}[{_ts()}]{C.RESET}  {C.CYAN}Test mode stopped.{C.RESET}")


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(
        description="Rahat Sentinel — Alert Receiver & BLE Encoder",
    )
    parser.add_argument(
        "--test", action="store_true",
        help="Skip WebSocket and run in simulated test mode",
    )
    args = parser.parse_args()

    if args.test:
        asyncio.run(_fallback_test_loop())
    else:
        asyncio.run(ws_listener())


if __name__ == "__main__":
    main()
